core/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -712,37 +710,55 @@ private static void parseCopy(String field, ParseContext context) throws IOExcep
+            String fieldName = paths[paths.length-1];
             ObjectMapper mapper = context.root();
-            String objectPath = "";
-            String fieldPath = field;
-            int posDot = field.lastIndexOf('.');
-            if (posDot > 0) {
-                objectPath = field.substring(0, posDot);
-                context.path().add(objectPath);
-                mapper = context.docMapper().objectMappers().get(objectPath);
-                fieldPath = field.substring(posDot + 1);
-            }
-            if (mapper == null) {
-                //TODO: Create an object dynamically?
-                throw new MapperParsingException("attempt to copy value to non-existing object [" + field + "]");
-            }
-            ObjectMapper update = parseDynamicValue(context, mapper, fieldPath, context.parser().currentToken());
-            assert update != null; // we are parsing a dynamic value so we necessarily created a new mapping
-
-            // propagate the update to the root
-            while (objectPath.length() > 0) {
-                String parentPath = "";
+            ObjectMapper[] mappers = new ObjectMapper[paths.length-1];
+            if (paths.length > 1) {
                 ObjectMapper parent = context.root();
-                posDot = objectPath.lastIndexOf('.');
-                if (posDot > 0) {
-                    parentPath = objectPath.substring(0, posDot);
-                    parent = context.docMapper().objectMappers().get(parentPath);
+                for (int i = 0; i < paths.length-1; i++) {
+                    context.path().add(paths[i]);
+                    mapper = context.docMapper().objectMappers().get(paths[i]);
+                    if (mapper == null) {
+                        // One mapping is missing, check if we are allowed to create a dynamic one.
+                        ObjectMapper.Dynamic dynamic = parent.dynamic();
+                        if (dynamic == null) {
+                            dynamic = dynamicOrDefault(context.root().dynamic());
+                        }
+                        if (dynamic == ObjectMapper.Dynamic.STRICT) {
+                            context.path().remove();
+                            throw new StrictDynamicMappingException(context.path().fullPathAsText(""), paths[i]);
+                        } else if (dynamic == ObjectMapper.Dynamic.TRUE) {
+                            context.path().remove();
+                            Mapper.Builder builder = context.root().findTemplateBuilder(context, paths[i], "object");
+                            if (builder == null) {
+                                builder = MapperBuilders.object(paths[i]).enabled(true).pathType(parent.pathType());
+                            }
+                            Mapper.BuilderContext builderContext = new Mapper.BuilderContext(context.indexSettings(), context.path());
+                            mapper = (ObjectMapper) builder.build(builderContext);
```

Comment:
> You want to disallow the creation of dynamic nested objects throw the copy_to ?

Exactly.

> if the user wants to force the creation of a nested objects for the copy_to target he just need to add the definition in the mapping

This belongs to a different issue, but I'm also wondering that copying to a nested field, even if it already exists should not be allowed as we have no idea in which nested doc the value should be copied?

To be clear, if you are currently parsing a.b.c and copying to a.b.d and a.b is a nested object field, it's fine since the source and target and under the same nested scope (so adding to the current nested doc that is being built is fine). However, if you're copying from a.b.c to a.d.e and a.d is nested, then we have no way to know to which nested document of a.d.e the value should go, so we should reject it?

(I'm starting this discussion here because your code made me think about this potential problem, but please don't try to address it in the current PR, I think this should go into a dedicated PR.)

> Bottom line (but maybe I am missing something) it seems weird to define a copy_to without defining the type of the target in the mapping (lazy users ? ;) ).

I tend to agree. I think the idea is that we already support dynamic field creation for fields, so we want to be consistent and do it with objects as well.